llama_index/agent/openai_agent.py

```diff
@@ -104,6 +138,66 @@ def chat(
 
         return Response(ai_message.content)
 
+    def stream_chat(
+        self, message: str, chat_history: Optional[List[ChatMessage]] = None
+    ) -> Generator[StreamingChatResponse, None, None]:
+        chat_history = chat_history or self._chat_history
+        chat_history.append(ChatMessage(content=message, role="user"))
+        tools = self._get_tools(message)
+        functions = [tool.metadata.to_openai_function() for tool in tools]
+
+        def _stream_chat(
```

Comment:
out of curiosity why is this a nested function

Reply:
Felt specific to this exact function so I wasn't sure if it would need to be accessible more broadly

Could move this to a util function, or a class method too. The code duplication thing is still an issue though 😢